proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1030 @@
+   ``-this-unit-id UnitId``, where ``UnitId`` consists of ``ComponentId``
+   and a generalized module substitution ``m=<m>``, for each ``m``
+   in the free module variables of ``dependency`` and each ``m``
+   in ``signature m``.
+
+A single module or signature in an indefinite unit can be
+typechecked with ``ghc -c`` by specifying only that module name.
+(this motivates the "redundant" specification of requirements in
+``this-unit-id``; it's not redundant when compiling with ``ghc -c``!)
+
+For compiling a definite unit with module substitution ``ModuleSubst``,
+the only difference is ``-this-unit-id`` now takes a ``UnitId``
+consisting of ``ComponentId`` and the specified substitution
+``ModuleSubst`` to instantiate the module with.  Note that the
+unit identifiers from dependencies do NOT have the substitution
+applied to them (GHC can read off the necesary substitution from
+``-this-unit-id``; in fact, applying this substitution would
+lose information.)
+
+Not all unit identifiers are valid as arguments to
+``this-unit-id``: only fully generalized unit identifiers
+(in the case of typechecking indefinite units) or unit identifiers
+with no free module variables (in the case of compiling a definite
+unit) are permissible.
+
+We impose an additional requirement that there must be an ``hsig`` file
+for every required signature of a unit (even if all nontrivial
+requirements are inherited from dependencies).  This is to preserve the
+compilation model of "one source file, one invocation of the compiler."
+
+Installed unit database
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The **installed unit database** (previously known as the installed
+package database) records the set of units which can be referenced by
+``dependency`` declarations.  A unit identifier uniquely identifies
+a unit in the installed unit database.
+
+Logically, entries in the installed unit database can be classified
+into three types:
+
+1. Non-Backpack libraries (units), for which the unit identifier is
+   equivalent to the component identifier,
+
+2. Definite Backpack units, where the unit identifier has a nontrivial
+   substition with no free module variables, and
+
+3. Indefinite Backpack units, where the unit identifier is in most
+   general form.
+
+Definite Backpack units and non-Backpack libraries never depend on
+indefinite Backpack units.  Indefinite Backpack units never refer
+to definite Backpack units (this makes typechecking possible even
+when no compilation has occurred at all); e.g., a dependency on a
+partially instantiated unit identifier refers to the indefinite
+unit identified by the generalized unit identifier.
+
+An entry in the installed unit database records a variety of
+information, but the most important for the purposes of Backpack
+are the **exposed modules**, which are a mapping from module names
+to module identities, which specify what modules get brought into
+scope when we declare a dependency on the library.
```

Comment:
You want to be able to say, here's a component id, give me the indefinite unit for it.